smaug/python/ops/control_flow_ops.py

```diff
@@ -83,7 +84,7 @@ def add_node(self, node, input_tensors):
     of its input tensor if it's deemed external to the context.
     """
     # Don't add the node if it's already added. Don't add a data op to avoid
-    # deadlock.
+    # an infinite loop.
```

Comment:
I finally see where the infinite loop is coming from, and that's the root of the problem. You should not be calling `context.add_node()` from `graph.add_node()` at all. Assuming that the context is necessary (and I still don't see why it is necessary), after having created the node with `graph.add_node()` (which gets called via `switch()`), just manually add that node to your internal context tracking structures. Then you don't need to expose `create_unique_name` or its really hacky `mark_as_used` parameter either.